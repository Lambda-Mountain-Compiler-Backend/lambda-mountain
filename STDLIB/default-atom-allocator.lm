
allocate-atom-counter := ();
allocate-atom-getcounter := λ. (tail(
   (if allocate-atom-counter () (
      (set allocate-atom-counter (uuid()))
   ))
   allocate-atom-counter
));
allocate-atom-count-prog := λ. (
   ( \t 'mov \s (allocate-atom-getcounter()) , \s '%r12 \n
     \t 'mov \s '0 \[ '%r12 \] , \s '%r12 \n
     \t 'mov \s '$0, \s '%r13 \n )
);

allocate-atom-grow-fid := ();
allocate-atom-grow := λ size . (tail(
   # INPUT:
   #    r8 : base pointer
   #    r9 : tail pointer
   # OUTPUT:
   #    r8  : base pointer (possibly modified)
   #    r9  : tail pointer (possibly modified)
   #    r10 : nothing      (possibly modified)
   (if allocate-atom-grow-fid () (
      (set allocate-atom-grow-fid (uuid()))

      (local page-size)
      (set page-size 1073741824)

      (local initialized)
      (set initialized (uuid()))

      (local assemble_atom_tail)
      (set assemble_atom_tail (uuid()))

      (set assemble-data-section ( assemble-data-section (
         assemble_atom_tail ':                       \n
         \t '.zero \s '8                             \n
      )))

      (set assemble-text-section ( assemble-text-section (
         allocate-atom-grow-fid ':                   \n
         \t 'push \s '%r10                           \n
         \t 'push \s '%rax                           \n
         \t 'push \s '%rdi                           \n
         \t 'cmp \s '$0, \s '%r8                     \n
         \t 'jne \s initialized                      \n
         \t 'mov \s '$ assemble_atom_tail , \s '%r10 \n
         \t 'mov \s '0 \[ '%r10 \] , \s '%r8         \n
         \t 'mov \s '%r8, \s '%r9                    \n
         \t 'cmp \s '$0, \s '%r8                     \n
         \t 'jne \s initialized                      \n
         \t 'mov \s '$12, \s '%rax                   \n  # system call brk
         \t 'mov \s '$0, \s '%rdi                    \n  # sbrk(0)
         \t 'syscall                                 \n
         \t 'mov \s '%rax, '%r8                      \n
         \t 'add \s '$ page-size , \s '%r8           \n
         \t 'add \s '$ page-size , \s '%r8           \n
         \t 'add \s '$ page-size , \s '%r8           \n
         \t 'add \s '$ page-size , \s '%r8           \n
         \t 'mov \s '$12, \s '%rax                   \n  # system call brk
         \t 'mov \s '%r8, \s '%rdi                   \n  # new requested brk
         \t 'syscall                                 \n
         \t 'sub \s '$ page-size , \s '%r8           \n
         \t 'sub \s '$ page-size , \s '%r8           \n
         \t 'sub \s '$ page-size , \s '%r8           \n
         \t 'sub \s '$ page-size , \s '%r8           \n
         \t 'mov \s '%r8, \s '%r9                    \n
         \t 'mov \s '%r9 , \s '%r10                  \n
         initialized ':                              \n
         \t 'mov \s '$ assemble_atom_tail , \s '%r10 \n
         \t 'add \s '%r11 , \s '%r9                  \n
         \t 'mov \s '%r9, \s '0 \[ '%r10 \]          \n
         \t 'sub \s '%r11 , \s '%r9                  \n
         \t 'pop \s '%rdi                            \n
         \t 'pop \s '%rax                            \n
         \t 'pop \s '%r10                            \n
         \t 'ret                                     \n
      )))
   ))

   (\t 'push \s '%r11                                \n
    \t 'mov \s size , '%r11                          \n
    \t 'call \s allocate-atom-grow-fid               \n
    \t 'pop \s '%r11                                 \n)
));
