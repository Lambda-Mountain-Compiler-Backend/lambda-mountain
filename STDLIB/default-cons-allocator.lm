
allocate-cons-s-section := ();
allocate-cons-s-counter := ();
allocate-cons := λ . (tail(
   (if allocate-cons-s-section () (
      (set allocate-cons-s-section (uuid()))
      (set allocate-cons-s-counter (uuid()))
      (set assemble-data-section ( assemble-data-section (
         allocate-cons-s-counter ': \n
         \t '.zero \s '8            \n
         allocate-cons-s-section ': \n
         \t '.zero \s '536870912    \n
      )))
   ))
   (\t 'push \s '%r10                                          \n
    \t 'push \s '%r11                                          \n
    \t 'mov \s '$ allocate-cons-s-section ', \s '%r8           \n  # %r8 now points to root s
    \t 'mov \s '$ allocate-cons-s-counter ', \s '%r11          \n  # %r11 now points to s counter
    \t 'mov \s '0 \[ '%r11 \] , \s '%r10                       \n  # %r10 now hold value of cons counter
    \t 'add \s '%r10, \s '%r8                                  \n  # %r8 now points to top free cons
    \t 'add \s '$16, \s '%r10                                  \n  # increment s counter
    \t 'mov \s '%r10, \s '0 \[ %r11 \]                         \n
    \t 'pop \s '%r11                                           \n
    \t 'pop \s '%r10                                           \n
   ) # overwrite new s counter
));

#allocate-cons-function-id := ();
#allocate-cons := λ . (tail(
#   (if allocate-cons-function-id () (
#      (set allocate-cons-function-id (uuid()))
#
#      (local extend)
#      (set extend (uuid()))
#      (local increment)
#      (set increment (uuid()))
#      (local initialized)
#      (set initialized (uuid()))
#      (local exit-allocation)
#      (set exit-allocation (uuid()))
#
#      (local page-size)
#      (set page-size 1073741824)
#
#      (set assemble-data-section ( assemble-data-section (
#         '__current_break:                     \n
#         \t '.zero \s '8                       \n
#         '__sequent_break:                     \n
#         \t '.zero \s '8                       \n
#      )))
#
#      (set assemble-text-section ( assemble-text-section (
#         allocate-cons-function-id ':                \n
#         \t 'push \s '%r9                            \n
#         \t 'mov \s '$__current_break, \s '%rdi      \n
#         \t 'mov \s '0 \[ '%rdi \] , \s '%r8         \n
#         \t 'cmp \s '$0, \s '%r8                     \n
#         \t 'jne \s initialized                      \n
#         \t 'mov \s '$__program_break, \s '%r8       \n
#         \t 'mov \s '%r8, \s '0 \[ '%rdi \]          \n
#         \t 'add \s '$ page-size , \s '%r8           \n
#         \t 'mov \s '$12, \s '%rax                   \n  # system call brk
#         \t 'mov \s '%r8, \s '%rdi                   \n  # new requested brk
#         \t 'syscall                                 \n
#         initialized ':                              \n
#         \t 'mov \s '$__current_break, \s '%r8       \n
#         \t 'mov \s '0 \[ '%r8 \] , \s '%r8          \n
#         \t 'add \s '$16, \s '%r8                    \n
#         \t 'mov \s '$__current_break, \s '%r9       \n
#         \t 'mov \s '%r8, \s '0 \[ '%r9 \]           \n
#         \t 'sub \s '$16, \s '%r8                    \n
#         \t 'pop \s '%r9                             \n
#         \t 'ret                                     \n
#      )))
#   ))
#   ( \t 'call \s allocate-cons-function-id \n )
#));
