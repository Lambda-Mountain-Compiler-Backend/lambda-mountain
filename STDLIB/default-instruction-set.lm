
type Reg8  AL  | BL  | CL  | DL  | SIL | DIL | BPL | SPL | R8B | R9B | R10B | R11B | R12B | R13B | R14B | R15B;
type Reg16 AX  | BX  | CX  | DX  | SI  | DI  | BP  | SP  | R8W | R9W | R10W | R11W | R12W | R13W | R14W | R15W;
type Reg32 EAX | EBX | ECX | EDX | ESI | EDI | EBP | ESP | R8D | R9D | R10D | R11D | R12D | R13D | R14D | R15D;
type Reg64 RAX | RBX | RCX | RDX | RSI | RDI | RBP | RSP | R8  | R9  | R10  | R11  | R12  | R13  | R14  | R15;

fragment Offset := λ(: o Offset). o;

fragment Reg8 := λ(: r Reg8). r;
fragment Reg16 := λ(: r Reg16). r;
fragment Reg32 := λ(: r Reg32). r;
fragment Reg64 := λ(: r Reg64). r;

fragment deref := λ(: r Reg64)(: dst Reg8). ( \t 'mov \s '0 \[ '% r \] , \s '% dst \n );
fragment deref := λ(: r Reg64)(: dst Reg16). ( \t 'mov \s '0 \[ '% r \] , \s '% dst \n );
fragment deref := λ(: r Reg64)(: dst Reg32). ( \t 'mov \s '0 \[ '% r \] , \s '% dst \n );
fragment deref := λ(: r Reg64)(: dst Reg64). ( \t 'mov \s '0 \[ '% r \] , \s '% dst \n );

fragment mov := λ(: imm U8)(: dst Reg8). ( \t 'mov \s '$ imm , \s '% dst \n );
fragment mov := λ(: imm U16)(: dst Reg16). ( \t 'mov \s '$ imm , \s '% dst \n );
fragment mov := λ(: imm U32)(: dst Reg32). ( \t 'mov \s '$ imm , \s '% dst \n );
fragment mov := λ(: imm U64)(: dst Reg64). ( \t 'mov \s '$ imm , \s '% dst \n );
fragment mov := λ(: imm I8)(: dst Reg8). ( \t 'mov \s '$ imm , \s '% dst \n );
fragment mov := λ(: imm I16)(: dst Reg16). ( \t 'mov \s '$ imm , \s '% dst \n );
fragment mov := λ(: imm I32)(: dst Reg32). ( \t 'mov \s '$ imm , \s '% dst \n );
fragment mov := λ(: imm I64)(: dst Reg64). ( \t 'mov \s '$ imm , \s '% dst \n );

fragment mov := λ(: src Reg8)(: dst Reg8). ( \t 'mov \s ''% src , \s '% dst \n );
fragment mov := λ(: src Reg16)(: dst Reg16). ( \t 'mov \s ''% src , \s '% dst \n );
fragment mov := λ(: src Reg32)(: dst Reg32). ( \t 'mov \s ''% src , \s '% dst \n );
fragment mov := λ(: src Reg64)(: dst Reg64). ( \t 'mov \s ''% src , \s '% dst \n );

fragment push := λ(: imm U8). ( \t 'pushb \s '$ imm \n );
fragment push := λ(: imm U16). ( \t 'pushw \s '$ imm \n );
fragment push := λ(: imm U32). ( \t 'pushs \s '$ imm \n );
fragment push := λ(: imm U64). ( \t 'pushq \s '$ imm \n );
fragment push := λ(: imm I8). ( \t 'pushb \s '$ imm \n );
fragment push := λ(: imm I16). ( \t 'pushw \s '$ imm \n );
fragment push := λ(: imm I32). ( \t 'pushs \s '$ imm \n );
fragment push := λ(: imm I64). ( \t 'pushq \s '$ imm \n );
fragment push := λ(: r Reg8). ( \t 'pushb \s '% r \n );
fragment push := λ(: r Reg16). ( \t 'pushw \s '% r \n );
fragment push := λ(: r Reg32). ( \t 'pushs \s '% r \n );
fragment push := λ(: r Reg64). ( \t 'pushq \s '% r \n );

fragment pop := λ(: r Reg8). ( \t 'popb \s '% r \n );
fragment pop := λ(: r Reg16). ( \t 'popw \s '% r \n );
fragment pop := λ(: r Reg32). ( \t 'pops \s '% r \n );
fragment pop := λ(: r Reg64). ( \t 'popq \s '% r \n );

fragment get := λ(: src Offset)(: dst Reg8). ( \t 'mov \s src \[ '%rbp \] , \s '% dst \n );
fragment get := λ(: src Offset)(: dst Reg16). ( \t 'mov \s src \[ '%rbp \] , \s '% dst \n );
fragment get := λ(: src Offset)(: dst Reg32). ( \t 'mov \s src \[ '%rbp \] , \s '% dst \n );
fragment get := λ(: src Offset)(: dst Reg64). ( \t 'mov \s src \[ '%rbp \] , \s '% dst \n );

fragment sub := λ(: imm U8)(: dst Reg8). ( \t 'sub \s $ imm \s , \s '% dst \n );
fragment sub := λ(: imm U16)(: dst Reg16). ( \t 'sub \s $ imm \s , \s '% dst \n );
fragment sub := λ(: imm U32)(: dst Reg32). ( \t 'sub \s $ imm \s , \s '% dst \n );
fragment sub := λ(: imm U64)(: dst Reg64). ( \t 'sub \s $ imm \s , \s '% dst \n );
fragment sub := λ(: imm I8)(: dst Reg8). ( \t 'sub \s $ imm \s , \s '% dst \n );
fragment sub := λ(: imm I16)(: dst Reg16). ( \t 'sub \s $ imm \s , \s '% dst \n );
fragment sub := λ(: imm I32)(: dst Reg32). ( \t 'sub \s $ imm \s , \s '% dst \n );
fragment sub := λ(: imm I64)(: dst Reg64). ( \t 'sub \s $ imm \s , \s '% dst \n );
fragment sub := λ(: l Reg8)(: dst Reg8). ( \t 'sub \s '% l \s , \s '% dst \n );
fragment sub := λ(: l Reg16)(: dst Reg16). ( \t 'sub \s '% l \s , \s '% dst \n );
fragment sub := λ(: l Reg32)(: dst Reg32). ( \t 'sub \s '% l \s , \s '% dst \n );
fragment sub := λ(: l Reg64)(: dst Reg64). ( \t 'sub \s '% l \s , \s '% dst \n );

fragment add := λ(: imm U8)(: dst Reg8). ( \t 'add \s $ imm \s , \s '% dst \n );
fragment add := λ(: imm U16)(: dst Reg16). ( \t 'add \s $ imm \s , \s '% dst \n );
fragment add := λ(: imm U32)(: dst Reg32). ( \t 'add \s $ imm \s , \s '% dst \n );
fragment add := λ(: imm U64)(: dst Reg64). ( \t 'add \s $ imm \s , \s '% dst \n );
fragment add := λ(: imm I8)(: dst Reg8). ( \t 'add \s $ imm \s , \s '% dst \n );
fragment add := λ(: imm I16)(: dst Reg16). ( \t 'add \s $ imm \s , \s '% dst \n );
fragment add := λ(: imm I32)(: dst Reg32). ( \t 'add \s $ imm \s , \s '% dst \n );
fragment add := λ(: imm I64)(: dst Reg64). ( \t 'add \s $ imm \s , \s '% dst \n );
fragment add := λ(: l Reg8)(: dst Reg8). ( \t 'add \s '% l \s , \s '% dst \n );
fragment add := λ(: l Reg16)(: dst Reg16). ( \t 'add \s '% l \s , \s '% dst \n );
fragment add := λ(: l Reg32)(: dst Reg32). ( \t 'add \s '% l \s , \s '% dst \n );
fragment add := λ(: l Reg64)(: dst Reg64). ( \t 'add \s '% l \s , \s '% dst \n );

fragment mul := λ(: imm U8)(: dst Reg8). ( \t 'mul \s $ imm \s , \s '% dst \n );
fragment mul := λ(: imm U16)(: dst Reg16). ( \t 'mul \s $ imm \s , \s '% dst \n );
fragment mul := λ(: imm U32)(: dst Reg32). ( \t 'mul \s $ imm \s , \s '% dst \n );
fragment mul := λ(: imm U64)(: dst Reg64). ( \t 'mul \s $ imm \s , \s '% dst \n );
fragment mul := λ(: imm I8)(: dst Reg8). ( \t 'mul \s $ imm \s , \s '% dst \n );
fragment mul := λ(: imm I16)(: dst Reg16). ( \t 'mul \s $ imm \s , \s '% dst \n );
fragment mul := λ(: imm I32)(: dst Reg32). ( \t 'mul \s $ imm \s , \s '% dst \n );
fragment mul := λ(: imm I64)(: dst Reg64). ( \t 'mul \s $ imm \s , \s '% dst \n );
fragment mul := λ(: l Reg8)(: dst Reg8). ( \t 'mul \s '% l \s , \s '% dst \n );
fragment mul := λ(: l Reg16)(: dst Reg16). ( \t 'mul \s '% l \s , \s '% dst \n );
fragment mul := λ(: l Reg32)(: dst Reg32). ( \t 'mul \s '% l \s , \s '% dst \n );
fragment mul := λ(: l Reg64)(: dst Reg64). ( \t 'mul \s '% l \s , \s '% dst \n );

fragment cmp := λ(: imm U8)(: dst Reg8). ( \t 'cmp \s '$ imm \s , \s '% dst \n );
fragment cmp := λ(: imm U16)(: dst Reg16). ( \t 'cmp \s '$ imm \s , \s '% dst \n );
fragment cmp := λ(: imm U32)(: dst Reg32). ( \t 'cmp \s '$ imm \s , \s '% dst \n );
fragment cmp := λ(: imm U64)(: dst Reg64). ( \t 'cmp \s '$ imm \s , \s '% dst \n );
fragment cmp := λ(: imm I8)(: dst Reg8). ( \t 'cmp \s '$ imm \s , \s '% dst \n );
fragment cmp := λ(: imm I16)(: dst Reg16). ( \t 'cmp \s '$ imm \s , \s '% dst \n );
fragment cmp := λ(: imm I32)(: dst Reg32). ( \t 'cmp \s '$ imm \s , \s '% dst \n );
fragment cmp := λ(: imm I64)(: dst Reg64). ( \t 'cmp \s '$ imm \s , \s '% dst \n );
fragment cmp := λ(: l Reg8)(: dst Reg8). ( \t 'cmp \s '% l \s , \s '% dst \n );
fragment cmp := λ(: l Reg16)(: dst Reg16). ( \t 'cmp \s '% l \s , \s '% dst \n );
fragment cmp := λ(: l Reg32)(: dst Reg32). ( \t 'cmp \s '% l \s , \s '% dst \n );
fragment cmp := λ(: l Reg64)(: dst Reg64). ( \t 'cmp \s '% l \s , \s '% dst \n );

fragment jmp := λ(: lbl Label). ( \t 'jmp \s lbl \n );
fragment je := λ(: lbl Label). ( \t 'je \s lbl \n );
fragment jne := λ(: lbl Label). ( \t 'jne \s lbl \n );
fragment jz := λ(: lbl Label). ( \t 'jz \s lbl \n );
fragment jnz := λ(: lbl Label). ( \t 'jne \s lbl \n );
fragment jlt := λ(: lbl Label). ( \t 'jlt \s lbl \n );
fragment jle := λ(: lbl Label). ( \t 'jle \s lbl \n );
fragment jgt := λ(: lbl Label). ( \t 'jgt \s lbl \n );
fragment jge := λ(: lbl Label). ( \t 'jge \s lbl \n );
