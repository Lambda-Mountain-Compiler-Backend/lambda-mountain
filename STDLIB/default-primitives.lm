
fragment type U8;     size U8 1;     atom suffix U8     _u8;
fragment type U16;    size U16 2;    atom suffix U16    _u16;
fragment type U32;    size U32 4;    atom suffix U32    _u32;
fragment type U64;    size U64 8;    atom suffix U64    _u64;
fragment type I8;     size I8 1;     atom suffix I8     _i8;
fragment type I16;    size I16 2;    atom suffix I16    _i16;
fragment type I32;    size I32 4;    atom suffix I32    _i32;
fragment type I64;    size I64 8;    atom suffix I64    _i64;
fragment type ASCII;  size ASCII 1;  atom suffix ASCII  _c;
fragment type String; size String 8; atom suffix String _s;

#[] := λ(: l Array<Sized<8>,?>)(: li U64). (: (
#   (mov( l R15 ))
#   (mov( r R14 ))
#   (shl( 3_u64 R14 ))
#   (add( R14 R15 ))
#   (mov( R15 0_u64 RAX ))
#   (as RAX x)
#) x);
#[] := λ(: l Array<Sized<4>,?>)(: li U64). (: (
#   (mov( l R15 ))
#   (mov( r R14 ))
#   (shl( 2_u64 R14 ))
#   (add( R14 R15 ))
#   (mov( R15 0_u64 RAX ))
#   (as RAX x)
#) x);
#[] := λ(: l Array<Sized<2>,?>)(: li U64). (: (
#   (mov( l R15 ))
#   (mov( r R14 ))
#   (shl( 1_u64 R14 ))
#   (add( R14 R15 ))
#   (mov( R15 0_u64 RAX ))
#   (as RAX x)
#) x);
#[] := λ(: l Array<Sized<1>,?>)(: li U64). (: (
#   (mov( l R15 ))
#   (mov( r R14 ))
#   (add( R14 R15 ))
#   (mov( R15 0_u64 RAX ))
#   (as RAX x)
#) x);

fragment [] := λ(: l LocalVariable+Array<x+Sized<1>,?>)(: r LocalVariable+U64). (: (
   (.program (
      \t 'mov \s (.expression l) \[ '%rbp \] , \s '%r15 \n
      \t 'mov \s (.expression r) \[ '%rbp \] , \s '%r14 \n
      \t 'add \s '%r14 , \s '%r15 \n
      \t 'mov \s '0 \[ '%r15 \] , \s '%r15b \n
   ))
   (.expression( 'r15b ))
) x+Reg8);


fragment [] := λ(: l LocalVariable+Array<x,?>)(: r GlobalVariable+U64). (: (
   (.program (
      \t 'mov \s '$ (.expression l) , \s '%r15 \n
      \t 'mov \s '$ (.expression r) , \s '%r14 \n
      \t 'mov \s '0 \[ '%r14 \] , \s '%r14 \n
      \t 'shl \s '$3 , \s '%r14 \n
      \t 'add \s '%r14 , \s '%r15 \n
      \t 'mov \s '0 \[ '%r15 \] , \s '%r15 \n
   ))
   (.expression( 'r15 ))
) x+Reg64);

fragment [] := λ(: l GlobalVariable+Array<x,?>)(: r GlobalVariable+U64). (: (
   (.program (
      \t 'mov \s '$ (.expression l) , \s '%r15 \n
      \t 'mov \s '$ (.expression r) , \s '%r14 \n
      \t 'mov \s '0 \[ '%r14 \] , \s '%r14 \n
      \t 'shl \s '$3 , \s '%r14 \n
      \t 'add \s '%r14 , \s '%r15 \n
      \t 'mov \s '0 \[ '%r15 \] , \s '%r15 \n
   ))
   (.expression( 'r15 ))
) x+Reg64);
fragment [] := λ(: l GlobalVariable+Array<x,?>)(: r LocalVariable+U64). (: (
   (.program (
      \t 'mov \s '$ (.expression l) , \s '%r15 \n
      \t 'mov \s (.expression r) \[ '%rbp \] , \s '%r14 \n
      \t 'shl \s '$3 , \s '%r14 \n
      \t 'add \s '%r14 , \s '%r15 \n
      \t 'mov \s '0 \[ '%r15 \] , \s '%r15 \n
   ))
   (.expression( 'r15 ))
) x+Reg64);
fragment [] := λ(: l GlobalVariable+Array<x,?>)(: r Constant+U64). (: (
   (.program (
      \t 'mov \s '$ (.expression l) , \s '%r15 \n
      \t 'mov \s '$ (.expression r) , \s '%r14 \n
      \t 'shl \s '$3 , \s '%r14 \n
      \t 'add \s '%r14 , \s '%r15 \n
      \t 'mov \s '0 \[ '%r15 \] , \s '%r15 \n
   ))
   (.expression( 'r15 ))
) x+Reg64);

fragment > := λ(: l GlobalVariable+U64)(: r Constant+U64). (: (
   (.program (
      \t 'mov \s '$ (.expression l) , \s '%r15 \n
      \t 'mov \s '0 \[ '%r15 \] , \s '%r15 \n
      \t 'cmp \s '$ (.expression r) , '%r15 \n
   ))
   (.expression( 'ja ))
) BranchConditional);
fragment > := λ(: l LocalVariable+U64)(: r Constant+U64). (: (
   (.program (
      \t 'mov \s (.expression l) \[ '%rbp \] , \s '%r15 \n
      \t 'cmp \s '$ (.expression r) , '%r15 \n
   ))
   (.expression( 'ja ))
) BranchConditional);

< := λ(: l U64)(: r U64). (: (tail(
   (gensym-label end)
   (mov( 1_u64 RAX ))
   (mov( l R8 ))
   (mov( r R9 ))
   (cmp( R9 R8 ))
   (jb end)
   (mov( 0_u64 RAX ))
   (label end)
   (as RAX U64)
)) U64);

+ := λ(: l U64)(: r U64). (: (tail(
   (mov( l R8 ))
   (mov( r RAX ))
   (add( R8 RAX ))
   (as RAX U64)
)) U64);

- := λ(: l U64)(: r U64). (: (tail(
   (mov( r R8 ))
   (mov( l RAX ))
   (sub( R8 RAX ))
   (as RAX U64)
)) U64);

% := λ(: l U64)(: r U64). (: (tail(
   (mov( 0_u64 RDX ))
   (mov( l RAX ))
   (mov( r R8 ))
   (div( R8 ))
   (as RDX U64)
)) U64);

/ := λ(: l U64)(: r U64). (: (tail(
   (mov( 0_u64 RDX ))
   (mov( l RAX ))
   (mov( r R8 ))
   (div( R8 ))
   (as RAX U64)
)) U64);

fragment branchtrue := λ(: _ Nil). (: (
   (.expression( 'jmp ))
) BranchConditional);
fragment branchfalse := λ(: _ Nil). (: (
   (.program( \t 'cmp \s '%rsp, \s '%rsp \n ))
   (.expression( 'jne ))
) BranchConditional);

!= := λ(: l U8)(: r U8). (: (tail(
   (push R14)
   (push R15)
   (gensym-label end-true)
   (gensym-label end-false)
   (gensym-label end)
   (mov( l R14B ))
   (mov( r R15B ))
   (cmp( R14B R15B ))
   (je end-false)
   (jmp end-true)
   (label end-true)
   (mov( 1_u64 RAX ))
   (jmp end)
   (label end-false)
   (mov( 0_u64 RAX ))
   (jmp end)
   (label end)
   (pop R15)
   (pop R14)
   (as RAX U64)
)) U64);

== := λ(: l U8)(: r U8). (: (tail(
   (push R14)
   (push R15)
   (gensym-label end-true)
   (gensym-label end-false)
   (gensym-label end)
   (mov( l R14B ))
   (mov( r R15B ))
   (cmp( R14B R15B ))
   (je end-true)
   (jmp end-false)
   (label end-true)
   (mov( 1_u64 RAX ))
   (jmp end)
   (label end-false)
   (mov( 0_u64 RAX ))
   (jmp end)
   (label end)
   (pop R15)
   (pop R14)
   (as RAX U64)
)) U64);
== := λ(: l U16)(: r U16). (: (tail(
   (push R14)
   (push R15)
   (gensym-label end-true)
   (gensym-label end-false)
   (gensym-label end)
   (mov( l R14W ))
   (mov( r R15W ))
   (cmp( R14W R15W ))
   (je end-true)
   (jmp end-false)
   (label end-true)
   (mov( 1_u64 RAX ))
   (jmp end)
   (label end-false)
   (mov( 0_u64 RAX ))
   (jmp end)
   (label end)
   (pop R15)
   (pop R14)
   (as RAX U64)
)) U64);
== := λ(: l U32)(: r U32). (: (tail(
   (push R14)
   (push R15)
   (gensym-label end-true)
   (gensym-label end-false)
   (gensym-label end)
   (mov( l R14D ))
   (mov( r R15D ))
   (cmp( R14D R15D ))
   (je end-true)
   (jmp end-false)
   (label end-true)
   (mov( 1_u64 RAX ))
   (jmp end)
   (label end-false)
   (mov( 0_u64 RAX ))
   (jmp end)
   (label end)
   (pop R15)
   (pop R14)
   (as RAX U64)
)) U64);
== := λ(: l U64)(: r U64). (: (tail(
   (push R14)
   (push R15)
   (gensym-label end-true)
   (gensym-label end-false)
   (gensym-label end)
   (mov( l R14 ))
   (mov( r R15 ))
   (cmp( R14 R15 ))
   (je end-true)
   (jmp end-false)
   (label end-true)
   (mov( 1_u64 RAX ))
   (jmp end)
   (label end-false)
   (mov( 0_u64 RAX ))
   (jmp end)
   (label end)
   (pop R15)
   (pop R14)
   (as RAX U64)
)) U64);


== := λ(: l I8)(: r I8). (: (tail(
   (push R14)
   (push R15)
   (gensym-label end-true)
   (gensym-label end-false)
   (gensym-label end)
   (mov( l R14B ))
   (mov( r R15B ))
   (cmp( R14B R15B ))
   (je end-true)
   (jmp end-false)
   (label end-true)
   (mov( 1_u64 RAX ))
   (jmp end)
   (label end-false)
   (mov( 0_u64 RAX ))
   (jmp end)
   (label end)
   (pop R15)
   (pop R14)
   (as RAX U64)
)) U64);
== := λ(: l I16)(: r I16). (: (tail(
   (push R14)
   (push R15)
   (gensym-label end-true)
   (gensym-label end-false)
   (gensym-label end)
   (mov( l R14W ))
   (mov( r R15W ))
   (cmp( R14W R15W ))
   (je end-true)
   (jmp end-false)
   (label end-true)
   (mov( 1_u64 RAX ))
   (jmp end)
   (label end-false)
   (mov( 0_u64 RAX ))
   (jmp end)
   (label end)
   (pop R15)
   (pop R14)
   (as RAX U64)
)) U64);
== := λ(: l I32)(: r I32). (: (tail(
   (push R14)
   (push R15)
   (gensym-label end-true)
   (gensym-label end-false)
   (gensym-label end)
   (mov( l R14D ))
   (mov( r R15D ))
   (cmp( R14D R15D ))
   (je end-true)
   (jmp end-false)
   (label end-true)
   (mov( 1_u64 RAX ))
   (jmp end)
   (label end-false)
   (mov( 0_u64 RAX ))
   (jmp end)
   (label end)
   (pop R15)
   (pop R14)
   (as RAX U64)
)) U64);
== := λ(: l I64)(: r I64). (: (tail(
   (push R14)
   (push R15)
   (gensym-label end-true)
   (gensym-label end-false)
   (gensym-label end)
   (mov( l R14 ))
   (mov( r R15 ))
   (cmp( R14 R15 ))
   (je end-true)
   (jmp end-false)
   (label end-true)
   (mov( 1_u64 RAX ))
   (jmp end)
   (label end-false)
   (mov( 0_u64 RAX ))
   (jmp end)
   (label end)
   (pop R15)
   (pop R14)
   (as RAX U64)
)) U64);

