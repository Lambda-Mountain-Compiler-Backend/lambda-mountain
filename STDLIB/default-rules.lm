
macro ( ('let x y) )
      ( (Î» x . ()) y );

macro ('match t ps) (tail(
   (let (uuid term) t)
   (let (uuid matched) ())
   (let (uuid return) ())
   (match-pats( (uuid term) (uuid matched) (uuid return) ps ))
   (uuid return)
));
macro ('match-pats( term matched return () )) ();
macro ('match-pats( term matched return (ps (lhs rhs)) )) (
   (match-pats( term matched return ps ))
   (if matched () (
      (set matched True)
      (match-destructure( term matched return lhs ))
      (if matched (
         (set return rhs)
      ) ())
   ))
);
macro ('match-destructure( term matched return () )) (
   (if matched (
      (if term (
         (set matched ())
      ) ())
   ) ())
);
macro ('match-destructure( term matched return (:Literal: l) )) (
   (if matched (
      (if (eq( term l )) () (
         (set matched ())
      ))
   ) ())
);
macro ('match-destructure( term matched return (:Variable: v) )) (
   (let v term)
);
macro ('match-destructure( term matched return (l r) )) (
   (let (uuid x) term)
   (set term (head (uuid x)))
   (match-destructure( term matched return l ))
   (set term (tail (uuid x)))
   (match-destructure( term matched return r ))
   (set term (uuid x))
);
