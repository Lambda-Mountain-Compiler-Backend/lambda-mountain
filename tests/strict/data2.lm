
import STDLIB/default-instruction-set.lm;
import STDLIB/default-primitives.lm;
import STDLIB/default-stdlib.lm;
import STDLIB/default-rules.lm;

type LR (Left( U8 , U16 , U32 , U64 )) | (Right( I64 , I64 , I32 , I16 , I8 ));

main := (
   (let l (Left( 1u8 2u16 3u32 4u64 )))
   (let r (Right( -4i64 -5i64 -6i32 -7i16 -8i8 )))
   (print (.0 l))
   (print (.1 l))
   (print (.2 l))
   (print (.3 l))
   (print (.4 l))
   (print (.0 r))
   (print (.1 r))
   (print (.2 r))
   (print (.3 r))
   (print (.4 r))
);
