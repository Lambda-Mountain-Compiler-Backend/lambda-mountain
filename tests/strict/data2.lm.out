143210-8-7-6-5
